middle_end/flambda/types/type_of_kind/type_of_kind_naked_immediate0.rec.ml

```diff
@@ -78,113 +78,187 @@ module Make_meet_or_join
+           a regular tag, so it's likely that the Naked_immediates set is more
+           relevant. *)
+        let tags =
+          I.Set.filter (fun tag ->
+              match Tag.create_from_targetint (Target_imm.to_targetint tag) with
+              | None -> false
+              | Some _ -> true)
+            tags
+        in
+        Ok (Naked_immediates tags, TEE.empty ())
       end
-    | (Is_int _ | Get_tag _), (Is_int _ | Get_tag _) -> Absorbing
+    | (Is_int _ | Get_tag _), (Is_int _ | Get_tag _) ->
+      (* We can't return Bottom, as it would be unsound, so we need to either
+         do the actual meet with Naked_immediates, or just give up and return
+         one of the arguments. *)
+      Ok (t1, TEE.empty ())
     | Naked_immediates tags, Get_tag ty ->
-      let tags =
-        Target_imm.Set.fold (fun tag tags ->
-            match Target_imm.to_tag tag with
-            | Some tag -> Tag.Set.add tag tags
-            | None -> bad_meet_or_join env t1 t2)
-          tags
-          Tag.Set.empty
-      in
-      begin match T.blocks_with_these_tags tags with
-      | Known shape ->
-        Or_bottom_or_absorbing.of_or_bottom
-          (E.switch T.meet T.join' env ty shape)
-          ~f:(fun (ty, env_extension) -> Get_tag ty, env_extension)
-      | Unknown ->
-        begin match E.op () with
-        | Meet -> Ok (Get_tag ty, TEE.empty ())
-        | Join ->
-          (* This should be Top, but we only have Absorbing; fortunately,
-             since we're in the join case, Absorbing turns out to be Top *)
-          Absorbing
+      let exception Invalid_tag in
+      begin try
+        let tags =
+          I.Set.fold (fun tag tags ->
+              match Target_imm.to_tag tag with
+              | Some tag -> Tag.Set.add tag tags
+              | None -> raise Invalid_tag)
+            tags
+            Tag.Set.empty
+        in
+        begin match T.blocks_with_these_tags tags with
+        | Known shape ->
+          Or_bottom.map
+            (T.meet env ty shape)
+            ~f:(fun (ty, env_extension) -> Get_tag ty, env_extension)
+        | Unknown ->
+          Ok (Get_tag ty, TEE.empty ())
         end
+      with Invalid_tag ->
+        (* There is a choice to make here: is it more interesting to keep
+           the Get_tag type, or the Naked_immediates type ?
+           In general the Get_tag type is more interesting, but we reach this
+           case only when we're doing a meet with something that cannot be
+           a regular tag, so it's likely that the Naked_immediates set is more
+           relevant. *)
+        let tags =
+          I.Set.filter (fun tag ->
+              match Tag.create_from_targetint (Target_imm.to_targetint tag) with
+              | None -> false
+              | Some _ -> true)
+            tags
+        in
+        Ok (Naked_immediates tags, TEE.empty ())
       end
+
+  let all_regular_tags =
```

Comment:
This should move to the Tag module.

Reply:
I forgot that `tag.mli` also exported a `Set` module. This should be fixed in the last commit.